File: 545. Boundary of Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def boundaryOfBinaryTree(self, root: TreeNode) -> List[int]:
        boundary = []
        # edge case
        if not root:
            return boundary
        boundary.append(root.val)
        
        # check left
        left = []
        node = root.left
        while node:
            left.append(node.val)
            if node.left:
                node = node.left
            else:
                node = node.right
                
        if left:
            left.pop()
        boundary.extend(left)
        
        # check bottom
        bottom = []
        stack = [root.right]
        node = root.left
        while node or stack:
            if node:
                if not node.left and not node.right:
                    bottom.append(node.val)
                    node = None
                else:
                    stack.append(node.right)
                    node = node.left
            else:
                node = stack.pop()
                
        boundary.extend(bottom)
        
        # check right
        right = []
        node = root.right
        while node:
            right.append(node.val)
            if node.right:
                node = node.right
            else:
                node = node.left
                
        if right:
            right.pop()
        right.reverse()
        boundary.extend(right)
        
        return boundary

File: 545. Boundary of Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        List<Integer> sol = new ArrayList<Integer>();
        if (root == null) {
            return sol;
        }
        sol.add(root.val);
        
        // left
        TreeNode node = root.left;
        while (node != null) {
            sol.add(node.val);
            if (node.left != null) 
                node = node.left;
            else
                node = node.right;
        }

        if (sol.size() > 1) 
            sol.remove(sol.size()-1);
        
        // bottom
        Stack<TreeNode> stack = new Stack<TreeNode>();
        node = root.left;
        stack.add(root.right);
        while (node != null || !stack.isEmpty()) {
            if (node != null) {
                if (node.left == null && node.right == null) {
                    sol.add(node.val);
                    node = null;
                } else {
                    stack.add(node.right);
                    node = node.left;
                }
            } else {
                node = stack.pop();
            }
        }
        
        // right
        List<Integer> aux = new ArrayList<Integer>();
        node = root.right;
        while (node != null) {
            aux.add(node.val);
            if (node.right != null) 
                node = node.right;
            else
                node = node.left;
        }

        if (aux.size() > 0)
            aux.remove(aux.size()-1);
        
        Collections.reverse(aux);
        sol.addAll(aux);
        
        return sol;
    }
}
